### PR TITLE
Feature/support helm in release manager

### PR DESCRIPTION
**Pull request by clemensutschig**
first cut of HELM support, which also once and for all eliviates the DeploymentConfig limitations

This is still dirty! - helm variables from component pipeline are NOT in the RM

### CODE REVIEW

**Comment by michaelsauter** [comment]
Alright ... I have silently watched this for a while ... first of, kudos to the nice work!

I thought about the Helm integration on a more general note, and I'm going to throw a bit of a curve ball to you ;)

My basic question is: do you even want to make the Helm integration work like the Tailor integration?

The Tailor integration is very "Deployment(Config) centric", because:

* that's what the first version of the shared lib did
* operators were not a thing
* it allowed easy implementation of some sanity checks and extraction of pod data to render the GxP docs

However, I think long-term you want to be in a place where you can basically deploy any Helm chart. This means that you do not know which resources are your deployment units (think CRDs from operators, like, for example, a "Cluster" resource for a Postgres cluster). So I am wondering if it wouldn't be nice to have the Helm "branch" somewhat separated from the Tailor part. In the Helm "branch", I would question which checks are really needed (e.g. the checks that Deployments are controlled by the release manager are probably not needed at all because the Helm chart cannot be exported from the DEV namespace, so by definition all things are version controlled). Or, is it really needed to collect pod data like IP addresses to put them into the GxP templates when IP addresses may change anyway? Maybe we can use other evidence instead which is better suited and easier to get?

So - just some food for thought. I am aware this would probably blow up the scope of this PR - but it may also make a few things easier that were difficult because of the current constraints?

File: docs/modules/jenkins-shared-library/partials/odsComponentStageCopyImage.adoc
Comment (clemensutschig) [inline docs/modules/jenkins-shared-library/partials/odsComponentStageCopyImage.adoc:16]: not needed

Comment (michaelsauter) [reply docs/modules/jenkins-shared-library/partials/odsComponentStageCopyImage.adoc:16]: Yes I think this is because src/org/ods/component/CopyImageOptions.groovy extends `Options`, which has the branch param. The docs are auto-generated from that. If you do not extend from `Options`, you should be able to get rid of it.

**Review by clemensutschig — COMMENTED** [review]
Martin - one big thing is the tests... we should write a bunch of tests all around this?

**Comment by michaelsauter** [comment]
From our conversation yesterday @clemensutschig:

* I really really like that this solution doesn't use the `latest` tag any longer, but instead the "short Git Sha tag". Awesome! Do you still want to maintain a `latest` tag at all? BTW - this will mean the spec is updated, which automatically triggers a rollout for deployments. There is some code that does a manual "rollout" if not automatically triggered if I am not mistaken - I think this is not necessary for the `Deployment` path ... you may want to check that.
* I really really like that it is not pausing / unpausing any longer, but I think that collides with #686. I do not really know the consequences of this. May need to document the differences in behaviour?
* From the current implementation, I do not see how one could supply different values files for different environments. So for example, one might maintain something like a `values.dev.yaml`, `values.test.yaml` and `values.prod.yaml` to configure the application depending on the environment. I'm linking https://github.com/opendevstack/ods-pipeline/blob/master/docs/tasks/ods-deploy-helm.adoc, which describes how this works for ODS Pipeline. The approach we took is quite similar to how Tailor can be configured differently based on environments / namespaces. Keep in mind that you also need the same logic for "secrets files" (see next bullet point), which are just encrypted values files.
* It looks like you still need to handle Helm "secrets files" in the "RM flow". In the component pipeline, the Helm upgrade is surrounded by code that exposes a secret key: https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy#L192. As a side note, we switched from GPG keys to AGE keys in ODS Pipeline because the handling / CLI tool is so much easier. But that's probably a topic for another day.
* I observed that the `version` / `appVersion` in the Helm chart is not connected to the "version" set in the release manager (typically from Jira). This may need documentation to avoid confusion? In ODS Pipeline we are overwriting the version of the Helm package with the one defined by the user in `ods.yaml`, but tbh I am not sure if that is worth it.
* You should add `--three-way-merge` to the `diff` invocation (see https://github.com/opendevstack/ods-pipeline/blob/master/deploy/ods-pipeline/charts/tasks/templates/task-ods-deploy-helm.yaml#L78-L81). While Helm 3 uses a 3-way merge during `upgrade`, helm-diff only compares previous/new by default. Adding the flag teaches helm-diff to use a 3-way merge as well, which is a good thing for consistency. Otherwise the diff may exit with 0, while the upgrade would have made a change.
* You also may want to set `HELM_DIFF_IGNORE_UNKNOWN_FLAGS=true` for the diff command (https://github.com/opendevstack/ods-pipeline/blob/master/cmd/deploy-with-helm/helm.go#L160) as that ensures the diff command also works when someone adds more flags to `helm upgrade` that helm-diff does not understand. If you add this env var, I believe you do not need https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/services/OpenShiftService.groovy#L140 anymore.

I hope this is helpful, let me know if there are other questions that I may help with!

**Comment by clemensutschig** [comment]
@michaelsauter - thanks for for the hard challenge . this helps a lot ..

`From the current implementation, I do not see how one could supply different values files for different environments. So for example, one might maintain something like a values.dev.yaml, values.test.yaml and values.prod.yaml to configure the application depending on the environment. I'm linking https://github.com/opendevstack/ods-pipeline/blob/master/docs/tasks/ods-deploy-helm.adoc, which describes how this works for ODS Pipeline. The approach we took is quite similar to how Tailor can be configured differently based on environments / namespaces` - is fixed now - thru a new param, which allows you to pass those env dependent files :)

File: src/org/ods/services/OpenShiftService.groovy
Comment (michaelsauter) [inline src/org/ods/services/OpenShiftService.groovy]: I do not understand this line. You wouldn't need to check since you handle the case that `--three-way-merge` is already given down below, no? Apart from that, `--three-way-merge` isn't a known flag for `helm upgrade`.

`--atomic` is filtered out because it is a known flag for `helm upgrade`, but it is not handled by `helm diff`. But as I said I think it's better to set the env var to ignore unknown flags in diff, otherwise you prevent most upgrade flags to be used as it would break helm diffing.

File: test/groovy/vars/OdsComponentStageRolloutOpenShiftDeploymentSpec.groovy
Comment (clemensutschig) [inline test/groovy/vars/OdsComponentStageRolloutOpenShiftDeploymentSpec.groovy:193]: I am surprised this was removed ... I would expect that we get a deployment back ... ?

File: vars/odsComponentStageCopyImage.groovy
Comment (clemensutschig) [inline vars/odsComponentStageCopyImage.groovy:9]: test missing ... and you can take a lot from the rolloutstage test above .. (here we should find imported images in the .builds)

File: CHANGELOG.md
Comment (clemensutschig) [inline CHANGELOG.md:4]: remove the orchestration .. restriction on WHAT? ..

File: src/org/ods/component/CopyImageStage.groovy
Comment (clemensutschig) [inline src/org/ods/component/CopyImageStage.groovy:49]: we should change this to returnStatus in the sh task ...(and add a nice label) -rather than the quick way as done here

File: src/org/ods/component/CopyImageStage.groovy
Comment (clemensutschig) [inline src/org/ods/component/CopyImageStage.groovy:88]: we should change this to contain the image name - and not just the component id ...

File: src/org/ods/component/CopyImageStage.groovy
Comment (michaelsauter) [inline src/org/ods/component/CopyImageStage.groovy:52]: `--src-tls-verify=false`?

File: src/org/ods/component/CopyImageStage.groovy
Comment (michaelsauter) [inline src/org/ods/component/CopyImageStage.groovy:56]: `--dest-tls-verify=false`?

File: src/org/ods/component/CopyImageStage.groovy
Comment (michaelsauter) [inline src/org/ods/component/CopyImageStage.groovy:60]: This try/catch seems useless ;)

File: src/org/ods/component/RolloutOpenShiftDeploymentOptions.groovy
Comment (michaelsauter) [inline src/org/ods/component/RolloutOpenShiftDeploymentOptions.groovy:53]: Is this meant to say that `env` gets replaced? If so, maybe change the notation to e.g. `.<env>.yml' - that would make it clearer for me.

Comment (michaelsauter) [reply src/org/ods/component/RolloutOpenShiftDeploymentOptions.groovy:53]: Forget my comment, understood how it works now. Maybe we can make an example here that tells users they need to create values.dev.yaml, values.test.yaml and values.prod.yaml - and then if they pass values.env.yaml here, those files will be picked up.

File: src/org/ods/component/CopyImageStage.groovy
Comment (michaelsauter) [reply src/org/ods/component/CopyImageStage.groovy]: Please see https://github.com/opendevstack/ods-pipeline/blob/master/cmd/deploy-with-helm/main.go#L189-L218, this should be possible.

**Comment by clemensutschig** [comment]
> > I think this iready to be merged .. @metmajer @michaelsauter - for a very big first cut - it's feature wise at what tailor does / did
> 
> **Thanks for all the help I received here.** Should I open a new issue to craft a quickstarter that uses helm so people see how it would work?

This warrants a bigger discussion- how we want to handle this ... :/ (e.g. do it for all quickstarters in one short, .. or a flag, .. or)

File: docs/modules/jenkins-shared-library/partials/odsComponentStageCopyImage.adoc
Comment (metmajer) [inline docs/modules/jenkins-shared-library/partials/odsComponentStageCopyImage.adoc:67]: Sure about this as an option?

File: docs/modules/jenkins-shared-library/partials/odsComponentStageRolloutOpenShiftDeployment.adoc
Comment (metmajer) [inline docs/modules/jenkins-shared-library/partials/odsComponentStageRolloutOpenShiftDeployment.adoc:177]: This might be problematic, since we rely on certain labels to able to correlate resources with their respective systems, e.g. with Aqua. FYI @maksymmaksym

File: src/org/ods/component/Context.groovy
Comment (metmajer) [inline src/org/ods/component/Context.groovy:77]: Would prefer to remove Docker since OpenShift serves a registry for OCI-compliant container images.

File: src/org/ods/component/CopyImageOptions.groovy
Comment (metmajer) [inline src/org/ods/component/CopyImageOptions.groovy:32]: Don't you want to use the `private` access modifier to make the internal nature transparent?

File: src/org/ods/component/CopyImageStage.groovy
Comment (metmajer) [reply src/org/ods/component/CopyImageStage.groovy:58]: @serverhorror if conversations have been resolved, please visibly "resolve" them.

File: src/org/ods/component/CopyImageStage.groovy
Comment (metmajer) [inline src/org/ods/component/CopyImageStage.groovy:62]: You have provided some inline comments above about the "algorithm". Why not continue here to aid readability?

File: src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy
Comment (metmajer) [inline src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy:139]: What happens if both Chart.yaml and the openshiftDir exists?

File: src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy
Comment (metmajer) [inline src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy:169]: Please put Tailor and Helm-related logic into a Strategy class and apply the one or the other strategy vs. having a single logic that tries to handle both. This is otherwise hard to read and mock/test.

Example: class Tailor {
    prepare(context,...)
    apply()
    isApplicable()
    ...
}

File: src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy
Comment (metmajer) [inline src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy:213]: Does `tailorApply` read `options.helmValues` here?

File: src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy
Comment (metmajer) [inline src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy:281]: See above, please provide a clean separation of concerns. Support and maintenance will be improved.

File: src/org/ods/orchestration/util/Project.groovy
Comment (metmajer) [inline src/org/ods/orchestration/util/Project.groovy:1531]: Why is this not part of a HelmService?

**Review by metmajer — CHANGES_REQUESTED** [review]
Hi @serverhorror @clemensutschig! I have some concerns about the quality of the code. There is little encapsulation and tests do not cover the many changes which were made. I am not confident approving this yet, code should show better cohesion and testability. Also, labels are not applied when using Helm, which will cause issues with Aqua. We had done extensive work previously and should seek to provide the same labels with all deployment methods.

**Comment by metmajer** [comment]
We had a discussion about dedicated Helm Quickstarters and decided that Helm/Tailor will be based on conventions of the existing Quickstarters. Helm is not a feature of a Quickstarter, it's just the deployment mechanism. What about the `fe-be-monorepo` quickstarter? Has this been tested?

**Comment by metmajer** [comment]
Sure, @serverhorror, let's have a chat on this 👍

**Comment by metmajer** [comment]
@maksymmaksym can you please post the data you currently monitor in Aqua to enable Incident Management?

**Comment by metmajer** [comment]
@serverhorror there are certain labels we need to apply to the deployed resources, answer from @maksymmaksym pending.

You asked for details: I suggest to get familiar with the *Strategy Pattern* to implement a deployment strategy. An abstract `DeploymentStrategy` class would implement logic which is generic and therefore independent from Tailor or Helm and provides *Template Methods* which need to be implemented by its concrete subclasses `TailorDeploymentStrategy` and `HelmDeploymentStrategy`. Template methods could focus on the initialisation, application, or post-processing (e.g. apply labels) of a particular deployment method.

The implementation of the strategy could start with taking the code I reviewed and extracting it as-is into the two concrete strategy classes, then remove what's not needed for this particular strategy. This way you could come up with a single `apply(context)` method per strategy, which does Helm or Tailor, but not both. Depending on whether Helm or Tailor would be used, the one of the other strategy would be applied. In the case of Helm, the `HelmDeploymentStrategy` might be served via `HelmService::getDeploymentStrategy()` method. A `HelmService` is useful for encapsulating additional service methods, e.g. `getHelmPrivateKeyCredentialsId()`. Once this is implemented, the contents of the concrete strategy classes could be refactored into additional template methods or methods in the abstract `DeploymentStrategy` class if the logic would otherwise be duplicated across the Helm and Tailor classes.

Does that help?

File: src/org/ods/component/CopyImageOptions.groovy
Comment (metmajer) [reply src/org/ods/component/CopyImageOptions.groovy:32]: @serverhorror You can remove the "X is an INTERNAL value" comments, since this is already indicated by the `private` access modifier. `protected` would have been an alternative if you need to make the same field available in a subclass, but that's not the case.

**Comment by maksymmaksym** [comment]
@metmajer for every deployed resource we would like to track the associated CI name and a person who deployed the workload.

**Comment by metmajer** [comment]
Thanks @maksymmaksym. Hmm.. but the name of the person who deployed is not contained in the current implementation. The end-responsibility of a deployed system is in the hands of the system lead who is documented elsewhere, correct? Can I assume that the CI name (System Name) would then be enough, albeit a bit less convenient?

**Comment by maksymmaksym** [comment]
@metmajer if there is a CIs in all the cases its should be fine.

**Comment by metmajer** [comment]
@serverhorror following the discussion with @maksymmaksym, the system name is enough. We do not want to add the user who deployed. Currently, we apply the following labels when using component pipelines or the Release Manager pipeline. `app.opendevstack.org/system-name` is the minimum requirement from security, but I urge you to apply the others from the Release Manager as well.

Screenshot 2022-08-18 at 22 22 08

**Comment by clemensutschig** [comment]
To which resource types does this apply?

Von meinem iPhone gesendet

**Comment by jafarre-viewnext** [comment]
> To which resource types does this apply? I guess we could try to find all resources and label them in one shot - rather than patching https://www.krenger.ch/blog/openshift-add-or-remove-label/#:~:text=You%20can%20find%20more%20information%20on%20labeling%20nodes,the%20example%20%E2%80%9Cmylabel%E2%80%9D%29%20from%20a%20node%20or%20pod%3A @serverhorror
> […](#)

The labels are currently applied to all resources in the following way:
- The quickstarter pipeline labels all created resources with default labels.
- The component pipeline updates the labels for every deployment, as some of the values may (and will) change.
- When the component pipeline updates the labels, it takes into account whether it is being run through the release manager and applies additional labels: system name, change id (project-version) and whether we are performing a developer preview or a release.
- It is also taken into account whether there is a helm chart, in order to apply helm-specific labels. This may need to be reviewed.
- The deployment configs are also updated to add/updated the labels to the templates. This allows new pods to be properly labeled whenever they are created.


`OpenShiftResourceMetadata` should do all the job without (or almost without) modification. @serverhorror if you see the need to modify it or just need more information about how it works, don't hesitate to contact me.

Note: it was a bit tricky to avoid unnecessary automatic rollouts while performing all these actions. It would be a good idea to make sure that, after the modifications, we are still performing the minimum possible number of rollouts (hopefully one).

File: src/org/ods/component/Context.groovy
Comment (clemensutschig) [reply src/org/ods/component/Context.groovy:77]: if we change this we need to change it in ods-core etx.. i suggest to leave for now

File: src/org/ods/orchestration/util/Project.groovy
Comment (clemensutschig) [reply src/org/ods/orchestration/util/Project.groovy:1531]: agree!

**Comment by metmajer** [comment]
@serverhorror besides what you‘re thinking, why specifically is the use of our existing functions a problem in conjunction with Helm and where specifically do you see a race condition?

**Comment by metmajer** [comment]
@serverhorror I thought that unwanted redeployments were previously addressed by @jafarre-bi. Juan, can you comment?

**Comment by metmajer** [comment]
Let me add a bit more context to your thinking @serverhorror.

1. We previously decided to attach metadata as OpenShift labels since downstream (required) security tools did not support consuming annotations.
2. Concerning the ownership of these metadata, we see ODS as an opinionated framework and therefore decided to attach specific metadata automatically and consistently for identification and housekeeping purposes. Therefore, it's also OK to override any user-provided input. We consider this viable since we use our own namespaces. Users are free to apply other labels or annotations as they see fit.

We had a discussion about the labelling strategy in https://github.com/opendevstack/ods-quickstarters/issues/512 and documented the final outcome in https://github.com/opendevstack/ods-jenkins-shared-library/pull/715. Hope this helps!

**Comment by metmajer** [comment]
@serverhorror @jafarre-bi given Martin's input that Pod-labels cannot be reliably applied using Kubernetes’ Deployment configurations (a standard with Helm charts), as opposed to OpenShift-specific DeploymentConfig resources, we're discussing internally how precisely we can enable security-related incident monitoring for the former without those labels in place with @maksymmaksym

**Comment by tbugfinder** [comment]
@serverhorror - Your awesome summary has two pro/con sections named "Mutating Admission Controller". Could you edit the post?

**Comment by metmajer** [comment]
@serverhorror thanks for the write-up! Currently travelling, but will comment shortly.

**Comment by metmajer** [comment]
@serverhorror will you consider @michaelsauter's PR for this one?

**Comment by metmajer** [comment]
Hello @jafarre-bi and @jorge-romero. After @maksymmaksym's approval, please give this contribution from @serverhorror a review. Martin has worked out a solution most welcome by our users and applied some decent refactorings. Please ensure refactorings are properly tested, after all deployment strategies are at the heart of delivery. Thanks much!

**Comment by metmajer** [comment]
@braisvq1996 once reviewed, I request that we include this feature in an ODS release soon after our current EDP release. We have users eagerly waiting for this.

File: src/org/ods/component/HelmDeploymentStrategy.groovy
Comment (jorge-romero) [inline src/org/ods/component/HelmDeploymentStrategy.groovy:109]: The way the variable refreshResources is used make the if condition in line 105 always be true, it does not depends on configuration or previous execution results

File: src/org/ods/component/HelmDeploymentStrategy.groovy
Comment (jafarre-bi) [inline src/org/ods/component/HelmDeploymentStrategy.groovy:132]: Null values are used to mark labels for deletion by OpenShiftResourceMetadata.updateMetadata(). In some cases, one or more optional labels may not apply anymore and need to be deleted. If you just ignore them, we may end up with wrong labels set.
Actually OpenShiftResourceMetadata.updateMetadata should do its job correctly, regardless of using Tailor or Helm. Is there any difference that makes it unsuitable for the Helm strategy? It should be possible to adapt it.

Comment (jafarre-bi) [reply src/org/ods/component/HelmDeploymentStrategy.groovy:132]: Sure, I will schedule a meeting

File: src/org/ods/component/CopyImageStage.groovy
Comment (metmajer) [inline src/org/ods/component/CopyImageStage.groovy:82]: Note to self and @jafarre-bi @jorge-romero for the planned activity of improving log messages, and error messages in particular, we may want to use class name prefixes instead of `!!!`. Also, any errors which are known and can be handled, we need to introduce error numbers which resolve to help documentation.

File: src/org/ods/component/HelmDeploymentStrategy.groovy
Comment (metmajer) [reply src/org/ods/component/HelmDeploymentStrategy.groovy:132]: @jafarre-bi were you able to resolve your questions in the meeting?

File: src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy
Comment (metmajer) [inline src/org/ods/component/RolloutOpenShiftDeploymentStage.groovy:18]: If this code and the imports on top are not used, can we remove it?

**Review by metmajer — COMMENTED** [review]
Hello @serverhorror, a few quick thoughts:

- Some Specs show commented-out code, is this code supposed to stay?
- There is no Spec available for the TailorDeploymentStrategy, but it would really beneficial to have one. We're operating at the core of delivery and I would appreciate your support here.

File: src/org/ods/component/HelmDeploymentStrategy.groovy
Comment (metmajer) [reply src/org/ods/component/HelmDeploymentStrategy.groovy:132]: @serverhorror ok, we'll discuss internally!

**Comment by metmajer** [comment]
> Is this not good enough?

@serverhorror Yes, it is for me. I missed that one!

File: src/org/ods/component/ScanWithAquaStage.groovy
Comment (metmajer) [reply src/org/ods/component/ScanWithAquaStage.groovy:237]: @serverhorror this concerns Aqua, and shouldn't be mixed into a Helm-related PR. Please undo and open a new PR with this change.

Comment (clemensutschig) [reply src/org/ods/component/ScanWithAquaStage.groovy:237]: This is a must fix - we saw this breakibg when Helm is used..

Comment (metmajer) [reply src/org/ods/component/ScanWithAquaStage.groovy:237]: @clemensutschig I clarified already with Martin. we will take a look to ensure that we're not inconsistent with other integrations that follow the same pattern.

Comment (hrcornejo) [reply src/org/ods/component/ScanWithAquaStage.groovy:237]: @metmajer @serverhorror we have now two integrations of this type: Aqua and SonarQube. The code afected for SonarQube is in: https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/component/ScanWithSonarStage.groovy#L282

Comment (jafarre-bi) [reply src/org/ods/component/ScanWithAquaStage.groovy:237]: The latest comment by @hrcornejo has already been solved. I consider this conversation resolved.

**Comment by jafarre-bi** [comment]
IMO this is ready to merge. I see there have been build errors that should be solved first.
I give my approval now, but please, do not merge it before the tests that are under development are complete and pass.
Thank you and congratulations, @serverhorror, for the terrific work!

**Review by jafarre-bi — APPROVED** [review]

**Review by metmajer — APPROVED** [review]
Well done, @serverhorror, kudos for the hard work and contribution :) Please wait with the merge until our test cases indicate success. We are working on it and will let you know soon. Then, let's please ensure we squash-merge this thing :)

**Comment by metmajer** [comment]
To add more context to this conversation and hopefully clarify: the PR by @serverhorror is most welcome and will be released early next year. Therefore, we are not under pressure of merging this code into `master`. We see benefits in making contributions, including our own ones, _Done_ before having them merged, which enables us to forget about them for now and expect successful test executions during our formal release process vs. being unnecessarily blocked later on when the deep knowledge has already started to vanish. Quality becomes of extra importance when building a Platform, especially when changing something at its core. Thanks for your understanding.

**Review by jafarre-bi — APPROVED** [review]

**Comment by metmajer** [comment]
Over the last weeks, @serverhorror @jafarre-bi and others worked on triaging random errors during pipeline executions using Helm. We didn't see any clear evidence that this was related to Helm directly but needed a better understanding of the overall stability to decide on the next steps. When the team was confident about the final result, we decided to consider the implementation Done and to give the merge a go. Thanks for everyone's patience. Some contributions need a close look to ensure the quality of the resulting feature. This was certainly the case here. Well done, everyone!

**Pull request merged**
